src/nn/transformer/decoder.rs

```diff
@@ -0,0 +1,233 @@
+use rand::Rng;
+
+use crate::prelude::*;
+
+/// A transformer decoder block. Different than the normal transformer block as this self attention accepts an additional sequence from the encoder.
+///
+/// # Generics
+/// - `M` The embedding size of token vectors from decoder.
+/// - `N` The embedding size of token vectors from encoder.
+/// - `K` The size of the keys in self attention.
+/// - `H` The number of attention heads.
+/// TODO: Doctests
+#[derive(Default, Debug)]
+pub struct TransformerDecoderBlock<
+    const M: usize,
+    const N: usize,
+    const I: usize,
+    const K: usize,
+    const H: usize,
+> where
+    Assert<{ M % H == 0 }>: ConstTrue,
+    Assert<{ K % H == 0 }>: ConstTrue,
+{
+    pub attn: MultiHeadAttention<M, N, K, M, H>,
+    pub ff: (
+        LayerNorm1D<M>,
+        Residual<(Linear<M, I>, ReLU, Linear<I, M>)>,
+        LayerNorm1D<M>,
+    ),
+}
+
+impl<const M: usize, const N: usize, const I: usize, const K: usize, const H: usize> ResetParams
+    for TransformerDecoderBlock<M, N, I, K, H>
+where
+    Assert<{ M % H == 0 }>: ConstTrue,
+    Assert<{ K % H == 0 }>: ConstTrue,
+{
+    fn reset_params<R: Rng>(&mut self, rng: &mut R) {
+        self.attn.reset_params(rng);
+        self.ff.reset_params(rng);
+    }
+}
+
+impl<const M: usize, const N: usize, const I: usize, const K: usize, const H: usize>
+    CanUpdateWithGradients for TransformerDecoderBlock<M, N, I, K, H>
+where
+    Assert<{ M % H == 0 }>: ConstTrue,
+    Assert<{ K % H == 0 }>: ConstTrue,
+{
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
+        self.attn.update(grads);
+        self.ff.update(grads);
+    }
+}
+
+impl<
+        const M: usize,
+        const N: usize,
+        const I: usize,
+        const K: usize,
+        const H: usize,
+        const S1: usize,
+        const S2: usize,
+        T: Tape,
+    > Module<(Tensor2D<S1, M, T>, Tensor2D<S2, N, T>)> for TransformerDecoderBlock<M, N, I, K, H>
+where
+    Assert<{ M % H == 0 }>: ConstTrue,
+    Assert<{ K % H == 0 }>: ConstTrue,
+    Assert<{ S1 * K == H * S1 * (K / H) }>: ConstTrue,
+    Assert<{ S2 * K == H * S2 * (K / H) }>: ConstTrue,
+    Assert<{ S2 * M == H * S2 * (M / H) }>: ConstTrue,
+    Assert<{ H * S1 * (M / H) == S1 * M }>: ConstTrue,
+{
+    type Output = (Tensor2D<S1, M, T>, T);
+
+    fn forward(&self, (input, from_enc): (Tensor2D<S1, M, T>, Tensor2D<S2, N, T>)) -> Self::Output {
+        let (input, input_tape) = input.split_tape();
+        let (x, enc_tape) = self
+            .attn
+            .forward((input.duplicate().put_tape(input_tape), from_enc));
+        (self.ff.forward(add(x, &input)), enc_tape)
```

Comment:
I think this should be `self.ff.forward(add(x.put_tape(enc_tape), &input)`, and then can remove the tape from the output tuple?